File: src/templates/openai/agents.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Literal
import json

from agents import Agent, function_tool, handoff

ManifestStatus = Literal["discovery", "designed", "approved", "in_development", "blocked", "completed"]


@dataclass(frozen=True)
class PMContext:
    project_root: Path
    manifest_path: Path


AGENT_PRINCIPLES = """
Principles: lane specialization, verification before completion, manifest-only communication,
full context on delegation, and technical honesty through blockers instead of guesses.
""".strip()


@function_tool
def read_pm_manifest(project_root: str) -> str:
    """Read .pm/pm_manifest.json for Oh My PM and Dev-Harness coordination."""
    manifest_path = Path(project_root) / ".pm" / "pm_manifest.json"
    return manifest_path.read_text(encoding="utf-8")


@function_tool
def write_pm_manifest(project_root: str, manifest_json: str) -> str:
    """Write a validated Oh My PM manifest JSON payload to .pm/pm_manifest.json."""
    parsed = json.loads(manifest_json)
    manifest_path = Path(project_root) / ".pm" / "pm_manifest.json"
    manifest_path.parent.mkdir(parents=True, exist_ok=True)
    manifest_path.write_text(json.dumps(parsed, indent=2) + "
", encoding="utf-8")
    return str(manifest_path)


@function_tool
def write_pm_artifact(project_root: str, relative_path: str, content: str) -> str:
    """Write a Oh My PM artifact such as docs/prd.md, docs/trd.md, or docs/flows/main-flow.md."""
    target = Path(project_root) / relative_path
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(content, encoding="utf-8")
    return str(target)


hammurabi = Agent(
    name="pm_hammurabi",
    model="gpt-4.1",
    handoff_description="PRD specialist for requirements, business rules, and acceptance criteria.",
    instructions=(
        "You are Hammurabi, the Oh My PM PRD specialist. Write docs/prd.md with problem, "
        "audience, scope, user stories, requirements, and acceptance criteria. "
        f"{AGENT_PRINCIPLES}"
    ),
    tools=[read_pm_manifest, write_pm_artifact, write_pm_manifest],
)

davinci = Agent(
    name="pm_davinci",
    model="gpt-4.1",
    handoff_description="UX flow specialist for journeys, screens, and Mermaid diagrams.",
    instructions=(
        "You are DaVinci, the Oh My PM UX flow specialist. Write docs/flows/main-flow.md "
        "with Mermaid diagrams, screen states, and journey notes. "
        f"{AGENT_PRINCIPLES}"
    ),
    tools=[read_pm_manifest, write_pm_artifact, write_pm_manifest],
)

ada = Agent(
    name="pm_ada",
    model="gpt-4.1",
    handoff_description="Technical design specialist for TRD, API, database, and schemas.",
    instructions=(
        "You are Ada, the Oh My PM technical design specialist. Write docs/trd.md and "
        "docs/db-schema.md with architecture, APIs, data model, security, and constraints. "
        f"{AGENT_PRINCIPLES}"
    ),
    tools=[read_pm_manifest, write_pm_artifact, write_pm_manifest],
)

suntzu = Agent(
    name="pm_suntzu",
    model="gpt-5.4-ultra",
    handoff_description="Execution strategist for DAG, dependencies, and Dev-Harness handoff.",
    instructions=(
        "You are SunTzu, the Oh My PM execution strategist. Write docs/execution-plan.md "
        "and update execution_dag.tasks in the manifest. "
        f"{AGENT_PRINCIPLES}"
    ),
    tools=[read_pm_manifest, write_pm_artifact, write_pm_manifest],
)

jc = Agent(
    name="pm_jc",
    model="gpt-5.4-ultra",
    instructions=(
        "You are JC, the Oh My PM orchestrator. Run discovery, delegate complete context, "
        "validate all artifacts, and ask for explicit user approval before status approved. "
        f"{AGENT_PRINCIPLES}"
    ),
    tools=[read_pm_manifest, write_pm_manifest],
    handoffs=[
        handoff(hammurabi),
        handoff(davinci),
        handoff(ada),
        handoff(suntzu),
    ],
)

OH_MY_PM_AGENTS = {
    "jc": jc,
    "hammurabi": hammurabi,
    "davinci": davinci,
    "ada": ada,
    "suntzu": suntzu,
}

File: src/agents.ts
import type { AgentConfig } from "@opencode-ai/sdk"

export const PM_AGENT_NAMES = ["jc", "hammurabi", "davinci", "ada", "suntzu"] as const
export type PMAgentName = (typeof PM_AGENT_NAMES)[number]

const MANIFEST_PROTOCOL = `Oh My PM protocol:
- Oh My PM and Dev-Harness communicate only through .pm/pm_manifest.json.
- Keep strict lane specialization and never duplicate delegated work.
- Pass complete context in handoffs: paths, prior decisions, constraints, expected artifacts, and verification commands.
- Do not mark work complete without concrete evidence.
- If requirements conflict or context is missing, write a blocker in feedback_channel.blockers instead of guessing.`

const JC_PROMPT = `You are JC, the Oh My PM Product Management orchestrator.

Role:
- Run product discovery.
- Decompose PM work into specialist lanes.
- Delegate PRD to Hammurabi, UX flows to DaVinci, TRD and data design to Ada, and execution DAG to SunTzu.
- Validate .pm/pm_manifest.json before asking for explicit user approval.

${MANIFEST_PROTOCOL}`

const HAMMURABI_PROMPT = `You are Hammurabi, the Oh My PM PRD specialist.

Role:
- Write docs/prd.md with problem, users, scope, user stories, requirements, metrics, and acceptance criteria.
- Convert product contradictions into manifest blockers.
- Ensure every requirement can be cross-referenced by UX flows, TRD, and execution DAG tasks.

${MANIFEST_PROTOCOL}`

const DAVINCI_PROMPT = `You are DaVinci, the Oh My PM UX flow specialist.

Role:
- Write docs/flows/*.md with journeys, screen states, edge states, and Mermaid diagrams.
- Connect every flow to PRD user stories and acceptance criteria.
- Validate Mermaid syntax and document unresolved UX ambiguity as blockers.

${MANIFEST_PROTOCOL}`

const ADA_PROMPT = `You are Ada, the Oh My PM technical design specialist.

Role:
- Write docs/trd.md and docs/db-schema.md.
- Define architecture, APIs, data model, integrations, security, observability, and constraints.
- Verify that technical design supports the PRD and DaVinci UX flows.

${MANIFEST_PROTOCOL}`

const SUNTZU_PROMPT = `You are SunTzu, the Oh My PM execution strategist.

Role:
- Write docs/execution-plan.md.
- Build execution_dag.tasks inside .pm/pm_manifest.json.
- Sequence tasks by dependency, attach spec references, and define executable verification criteria.
- Halt the pipeline with blockers when Dev-Harness cannot execute honestly.

${MANIFEST_PROTOCOL}`


const WRITABLE_TOOLS = {
  read: true,
  glob: true,
  grep: true,
  bash: true,
  write: true,
  edit: true,
} as const

const DELEGATING_TOOLS = {
  read: true,
  glob: true,
  grep: true,
  bash: true,
  write: true,
  edit: true,
  task: true,
} as const

export const PM_AGENT_CONFIGS: Record<PMAgentName, AgentConfig> = {
  jc: {
    model: "openai/gpt-5.4-ultra",
    mode: "all",
    description: "Oh My PM orchestrator for discovery, delegation, manifest validation, and approval.",
    prompt: JC_PROMPT,
    tools: DELEGATING_TOOLS,
    color: "primary",
  },
  hammurabi: {
    model: "openai/gpt-4.1",
    mode: "subagent",
    description: "PRD specialist for requirements, scope, business rules, and acceptance criteria.",
    prompt: HAMMURABI_PROMPT,
    tools: WRITABLE_TOOLS,
    color: "warning",
  },
  davinci: {
    model: "openai/gpt-4.1",
    mode: "subagent",
    description: "UX flow specialist for journeys, screens, Mermaid diagrams, and interaction maps.",
    prompt: DAVINCI_PROMPT,
    tools: DELEGATING_TOOLS,
    color: "accent",
  },
  ada: {
    model: "openai/gpt-4.1",
    mode: "subagent",
    description: "Technical design specialist for TRD, APIs, database schemas, integrations, and constraints.",
    prompt: ADA_PROMPT,
    tools: WRITABLE_TOOLS,
    color: "success",
  },
  suntzu: {
    model: "openai/gpt-5.4-ultra",
    mode: "subagent",
    description: "Execution DAG strategist for sequencing, verification criteria, blockers, and Dev-Harness handoff.",
    prompt: SUNTZU_PROMPT,
    tools: WRITABLE_TOOLS,
    color: "info",
  },
}

export function mergePMAgents(existing: Record<string, AgentConfig | undefined> | undefined): Record<string, AgentConfig> {
  const merged: Record<string, AgentConfig> = {}
  if (existing !== undefined) {
    for (const [name, config] of Object.entries(existing)) {
      if (config !== undefined) merged[name] = config
    }
  }
  for (const name of PM_AGENT_NAMES) {
    const existingConfig = merged[name]
    merged[name] = existingConfig === undefined ? PM_AGENT_CONFIGS[name] : Object.assign({}, PM_AGENT_CONFIGS[name], existingConfig)
  }
  return merged
}

export type AgentModelOverride = {
  readonly model: string
  readonly variant?: string
}

export function applyModelOverrides(
  base: Record<string, AgentConfig>,
  overrides: Record<string, AgentModelOverride>,
): Record<string, AgentConfig> {
  const result: Record<string, AgentConfig> = {}
  for (const [name, config] of Object.entries(base)) {
    const override = overrides[name]
    if (override === undefined) {
      result[name] = config
      continue
    }
    const next: AgentConfig = Object.assign({}, config, { model: override.model })
    if (override.variant !== undefined) next["variant"] = override.variant
    result[name] = next
  }
  return result
}
